File: sl/common/socket.py
import asyncio
import inspect

from common import Logger
from common import utils

class Socket():
    source = None
    owner = None

    __await_events = None
    __events_results = None
    def __init__(self, source=None, owner=None):
        self.source = source
        self.owner = owner
        if source is not None:
            source.connect(self)

        self.__await_events = dict()
        self.__events_results = dict()

    async def push(self, data):
        try:
            await self.source.push(source=self, data=data)
        except Exception as e:
            Logger.log_error(e)

    async def close(self):
        self.source.disconnect(self)

    def set_owner(self, owner):
        self.owner = owner

    async def on_data(self, data): pass

    async def _data_recieved(self, data):
        try:
            action_result = await self.__assume_action(data)
            if data.get('type', None) == 'service':
                Logger.log_info('data_recieved')
                Logger.log_info(self.__await_events.keys())
            if action_result is not None:
                await self.push(data=action_result)
            elif (isinstance(data, dict)
                    and data.get('type', None) == 'service'
                    and data['id'] in self.__await_events.keys()):
                pd_item = utils.dict_to_pandas(
                        data['action_result'])
                result = (pd_item if pd_item is not None
                        else data['action_result'])
                self.__events_results[data['id']] = result
                self.__await_events[data['id']].set()
            else:
                await self.on_data(data)
        except Exception as e:
            Logger.log_error(e)

    async def __assume_action(self, data):
        response = None

        try:
            if (self.owner is not None
                    and data.get('type', None) == 'service'
                    and data.get('action', None) is not None
                    and data['action'].split('.')[0] == self.owner.name):

                if not hasattr(self.owner, data['action'].split('.')[1]):
                    raise Warning('%s have not action %s'%(
                            data['action'].split('.')[0],
                            data['action'].split('.')[1],))

                Logger.log_info('accepted_action')

                action = getattr(self.owner, data['action'].split('.')[1])
                if inspect.iscoroutinefunction(action):
                    action_result = await action(
                            *data.get('args', []),
                            **data.get('kwargs', {}),)
                elif inspect.ismethod(action):
                    action_result = action(
                            *data.get('args', []),
                            **data.get('kwargs', {}),)
                else:
                    action_result = action

                Logger.log_info('action produced')
                response = {
                        'type':'service',
                        'id':data['id'],
                        'action_result':action_result,}
                Logger.log_info(response)
        except Exception as e:
            response = {
                    'type':'service',
                    'id':data['id'],
                    'action_result':{'error':str(e)},}
        finally:

            return response

    async def execute(self, action, *args, **kwargs):
        result = None

        try:
            nonce = utils.get_nonce()
            Logger.log_info('execute')
            Logger.log_info(nonce)
            self.__await_events[nonce] = asyncio.Event()
            await self.push({
                'type':'service',
                'id': nonce,
                'action':action,
                'args':args,
                'kwargs':kwargs,})

            await self.__await_events[nonce].wait()
            Logger.log_info('executed')
            result = self.__events_results[nonce]
            Logger.log_info(result)
            del self.__events_results[nonce]
            del self.__await_events[nonce]
        except Exception as e:
            Logger.log_error(e)

        finally:
            return result

File: sl/sv/static/js/socket.js
export default class {
    constructor({source=undefined, owner=undefined}){
        this.source = source;
        this.owner = owner;

        if (source){
            source.connect(this);}}

    async push(data){
        await this.source.push({source=this, data=data});}

    close(){
        this.source.disconnect(this);}

    set_owner(owner){
        this.owner = owner;}

    on_data(data){}

    async _data_recieved(data){
        var action_result;
        action_result = await this.__assume_action(data);

        if (action_result){
            await this.push(action_result);}
        else if (data instanceof Object
                && data['type'] === 'service'
                && this.__await_events[data['id']]){
            this.__events_results[data['id']] = data['action_result'];
            this.__await_events[data['id']](data['action_result']);}
        else {
            await this.on_data(data);}}

    async __assume_action(data){
        var response, action, action_result, args;

        try {
            if (this.owner
                    && data['type'] === 'service'
                    && data['action']
                    && data['action'].split('.')[0] === this.owner.name){
                if(this.owner.hasOwnProperty(data['action'].split('.')[1])){
                    action = this.owner[data['action'].split('.')[1]];
                    if (action instanceof Function){
                        args = data['args'];
                        if (Object.keys(data['kwargs']).length > 0){
                            args.append(data['kwargs']);}

                        if (action.constructor.name === 'AsyncFunction'){
                            action_result = await action.apply(
                                    this.owner,
                                    args,);}
                        else {
                            action_result = action(
                                    this.owner,
                                    data['args'],);}}
                    else {
                        action_result = action;}}
                else {
                    throw new Error(`${data['action'].split('.')[0]}`
                        +` has no method ${data['action'].split('.')[1]}`);}}}
        catch(e) {
            action_result = {'error':e.toString()};}
        finally {
            if (action_result !== undefined){
                response = {
                        'type':'service',
                        'id':data['id'],
                        'action_result':action_result,};}}
        return response;}

    async execute(action, args, kwargs){
        var self = this;
        var nonce = new Date().getTime().toString();
        var result;

        await self.push({
            'type':'service',
            'id': nonce,
            'action':action,
            'args':args,
            'kwargs':kwargs,})

        await new Promise((resolve, reject)=>{
            self.__await_events[nonce] = resolve;})

        result = this.__events_results[nonce];

        delete this.__events_results[nonce];
        delete this.__await_events[nonce];

        return result;}}
